### ISSUE TITLE
ModuleDependency.getConfiguration() deprecation warning from Gradle internal code

### ISSUE THREAD

**Issue by DanielThomas** [gradle/gradle#1311]
Deprecation trace indicates this is a Gradle internal call:

	ModuleDependency.getConfiguration() has been deprecated and is scheduled to be removed in Gradle 4.0. Use ModuleDependency.getTargetConfiguration() instead.
	 	at org.gradle.api.internal.artifacts.dependencies.AbstractModuleDependency.getConfiguration(AbstractModuleDependency.java:56)
	 	at org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependency_Decorated.getConfiguration(Unknown Source)
	 	at org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependency.getProjectConfiguration(DefaultProjectDependency.java:75)
	 	at org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependency_Decorated.getProjectConfiguration(Unknown Source)
	 	at netflix.nebula.dependency.recommender.DependencyRecommendationsPlugin.applyRecommendationToDependency(DependencyRecommendationsPlugin.java:37)

### Your Environment
Gradle 3.4-rc-1

**Comment by bmuschko** [comment]
I can reproduce the scenario with the following test project:

```
subprojects {
    apply plugin: 'java'
}

project('a')

project('b') {
    dependencies {
        compile project(':a')
    }
    
    configurations.compile.incoming.dependencies.each { dependency ->
        if (dependency instanceof ProjectDependency) {
            println dependency.projectConfiguration
        }
    }
}
```

The build actually renders two different deprecations:

```
The ProjectDependency.getProjectConfiguration() method has been deprecated and is scheduled to be removed in Gradle 4.0.
        at build_agkuf78oyhbpa28zyklc3l7ws$_run_closure2$_closure4.doCall(/Users/bmuschko/dev/projects/gradle-playground/issue-1311/build.gradle:14)
ModuleDependency.getConfiguration() has been deprecated and is scheduled to be removed in Gradle 4.0. Use ModuleDependency.getTargetConfiguration() instead.
        at build_agkuf78oyhbpa28zyklc3l7ws$_run_closure2$_closure4.doCall(/Users/bmuschko/dev/projects/gradle-playground/issue-1311/build.gradle:14)
```

Here's the reason why you are seeing this:

The [code in your plugin](https://github.com/nebula-plugins/nebula-dependency-recommender-plugin/blob/master/src/main/groovy/netflix/nebula/dependency/recommender/DependencyRecommendationsPlugin.java#L37) uses the deprecated method [`ProjectDependency.getProjectConfiguration()`](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ProjectDependency.html#getProjectConfiguration()). You can see that the method renders the [first deprecation message](https://github.com/gradle/gradle/blob/release/subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultProjectDependency.java#L74). Internally, the method also calls [`ModuleDependency.getConfiguration()`](https://github.com/gradle/gradle/blob/release/subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultProjectDependency.java#L75) which is [deprecated as well](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ModuleDependency.html#getConfiguration()). It seems like we didn't switch to `getTargetConfiguration()` internally. That's a miss on our part.

To fix the issue, simply call `ProjectDependency.getTargetConfiguration()`  instead of `ProjectDependency.getProjectConfiguration()`. Given that the deprecated methods will be removed with 4.0 (potentially the next Gradle version anyway), I don't think it makes much sense to change the code internally in Gradle. I will take action if we realize that Gradle 4.0 will be further out than expected.

**Comment by DanielThomas** [comment]
Thanks @bmuschko. Will do!

**Comment by DanielThomas** [comment]
I'll add this here for others that run into this, because the ergonomics of the new API weren't immediately obvious:

```
Configuration configuration;
                try {
                    ProjectDependency.class.getMethod("getTargetConfiguration");
                    String targetConfiguration = projectDependency.getTargetConfiguration() == null ? Dependency.DEFAULT_CONFIGURATION : projectDependency.getTargetConfiguration();
                    configuration = projectDependency.getDependencyProject().getConfigurations().getByName(targetConfiguration);
                } catch (NoSuchMethodException e) {
                    //noinspection deprecation
                }
```

**Comment by wilkinsona** [comment]
FWIW, for the sake of plugin developers, I think this should have been fixed by changing Gradle. Without that change in a single codebase, you're asking many plugin developers to introduce the reflective hackery that @DanielThomas has shared above. That seems inefficient when considering the whole Gradle ecosystem

**Comment by bmuschko** [comment]
@DanielThomas @wilkinsona Thanks for the additional feedback. There are two aspects you mentioned.

1. **The changed behavior of the new API:** The new API doesn't just replace the method. It actually changes the runtime behavior. As a result, it's forces the end user into having to implement code for handling a possible `null` value as shown in your code example. I am not quite sure why this design decision was made but maybe @adam might be able to comment on it. From my perspective it would be more convenient for end users if the method would just return the `default` configuration if no matching configuration could be resolved.
2. **Backwards compatibility:** We usually give end users a full major release to adapt their plugins from a deprecated API to a replacement API. The timeframe is roughly 1 year for switching over. I think that's plenty of time. In this specific case we deprecated the method with 3.0 and will remove it with 4.0. You'd only need to introduce reflection here _if_ you wanted your plugin to still support Gradle 2.x versions after the old API has been removed (>= 4.0). 

Here's my personal opinion about the topic from two different perspectives:

1. **View from a Gradle core developer's perspective:** I do understand your concern. You want to build a plugin with 3.x which should still be compatible with consumers of Gradle 2.x and doesn't render a deprecation warning. From my perspective having this kind of code in Gradle core would significantly impact the maintainability of the code base. We'd have to put in the handling for pretty much any method that has been deprecated in the past. So far the Gradle team did not adopt the practice and I doubt we'll do so in the future. Either way I am going to bring it up internally for discussion.
2. **View from a plugin developer's perspective:** Before working as a Gradle core developer I have been implementing and maintaining external plugins for years (easily 10-15 plugins). My rule of thumb in the past has been: I am going to have to use new APIs in the future instead of deprecated APIs. If that's the case I will release a new version of the plugin that requires a minimal Gradle version to work properly. Any consumer of my plugin that would like to use new features of my plugin will have to upgrade to a new version of Gradle. I usually didn't back port features to patch releases. The important point is to document the Gradle version that is compatible with a specific version of my plugin. My assumption was that users will want to upgrade to new Gradle versions over time as they provide many new features and performance improvements. 
Granted we are working in a very conservative domain. Enterprises are often reluctant to upgrading their build system as it might have a (potentially negative) impact on developer productivity. Thinking back, I can't really remember a single consumer of my plugin that complained that he/she wasn't willing to upgrade just because my plugin didn't support a old Gradle version. Maybe my plugins have been less popular than yours or were used by a different user base but I must say I have not run into a lot of issues on that front. Disclaimer: My argument doesn't touch on the aspect of binary compatibility. That's a different story.
BTW: Optimally, the plugin portal should run automated functional tests for published plugin versions against a range of Gradle versions to indicate backwards and forwards compatibility. It's a feature I'd personally like to see. Unfortunately, it didn't stack up against other priorities yet.

I hope my comment is helpful. Keep in mind that it reflects my personal opinion on the topic. Other Gradle engineers might see it differently. Please let me know if you want to discuss this further in more detail.

**Comment by wilkinsona** [comment]
@bmuschko Thanks for sharing your opinion on this. It's certainly helpful. 

> You'd only need to introduce reflection here if you wanted your plugin to still support Gradle 2.x versions after the old API has been removed

That's not the case. As things stand, we need to use reflection if we want to support 2.x and 3.x without a deprecation warning that our users are complaining about.

> You want to build a plugin with 3.x which should still be compatible with consumers of Gradle 2.x and doesn't render a deprecation warning

Exactly.

> From my perspective having this kind of code in Gradle core would significantly impact the maintainability of the code base

The alternative is that every plugin developer who wants to support 2.x and 3.x has to impact the maintainability of their codebase by introducing reflective hacks. As I said above, I think that's an inefficient way to tackle the problem and I would much rather see Gradle take care of this for everyone.

> The important point is to document the Gradle version that is compatible with a specific version of my plugin

This is what we do. For Spring Boot 1.5.x, we state that we support Gradle 2.x (2.9+) and 3.x. We can't drop support for 2.x in a 1.5.x release of Spring Boot to get rid of a deprecation warning. Having to move to a new major version of Gradle to use a new maintenance release of Spring Boot would be unacceptable to our users.

It's perhaps a little extreme, but in terms of usability I think it could be argued that Gradle 3.4 is not backwards compatible with 3.3. A deprecation warning is now shown (which users quite rightly want to get rid of) and there's nothing they can do about it.

**Comment by melix** [comment]
The rationale behind this change is simple: before, we were not able to make the difference between a dependency that falls back to the default configuration, and an explicit dependency on the `default` configuration:

```
dependencies {
    compile project(':foo') // this will fallback on the `default` configuration
}
```
and 
```
dependencies {
    compile project(':foo', configuration:'default') // this is an explicit dependency on the `default` configuration
}
```

The dependency engine needs to make the difference between the two cases. In the first one, we only set a dependency onto a project, and its the responsibility of Gradle to determine what target configuration to use. It can be `default`, but it can be something else. In particular, now that we support variant aware dependency management, the configuration that is going to be used depend on the consumer.

So, in short, the deprecation warnings that you see are legitimate. You should not call `ProjectDependency#getProjectConfiguration` because it will not reflect what the dependency engine will really do (there's no longer a one to one relation between a project dependency and a target configuration). The method is still here precisely for backwards compatibility, but if we don't show a deprecation message, you, as plugin authors, would not realize you are calling a method that doesn't make sense anymore.

**Comment by oehme** [comment]
Instead of using reflection you could also check `project.getGradle().getGradleVersion()`.

> As I said above, I think that's an inefficient way to tackle the problem and I would much rather see Gradle take care of this for everyone.

Could you elaborate how Gradle could do that? There is a plugin that is calling a deprecated method. We need to let the user know that this will no longer work in the next major release. The only way I currently see is for plugin authors to have two code paths if they want to stay backwards compatible after a deprecation. 

Making deprecations a "breaking change" would seriously lock down our options and not really improve the situation. You'd still have only two options: Support the previous major version and add two code paths or drop support for the previous major version immediately.

> The new API doesn't just replace the method. It actually changes the runtime behavior. As a result, it's forces the end user into having to implement code for handling a possible null value as shown in your code example. I am not quite sure why this design decision was made but maybe @adam might be able to comment on it. From my perspective it would be more convenient for end users if the method would just return the default configuration if no matching configuration could be resolved.

This is not a resolved dependency, but a *declared* dependency. We don't know yet what it will resolve to. Gradle has variant aware dependency resolution now, so a project dependency can resolve to different Configurations depending on which attributes the resolving Configuration has. Thus returning `default` would be wrong. 

Without knowing the exact use case of the call, it is hard to give a suggestion. But the root of the issue is that @DanielThomas' code is currently assuming that it knows what a project dependency will resolve to, which is no longer true in Gradle 3.4

**Comment by wilkinsona** [comment]
> The method is still here precisely for backwards compatibility, but if we don't show a deprecation message, you, as plugin authors, would not realize you are calling a method that doesn't make sense anymore.

The problem for me really boils down to how that deprecation warning is being shown. Users are being told about a deprecation that is (largely) outside of their control to fix. Plugin authors can realise that things are deprecated by compiling against the latest API and failing on warnings.

> Instead of using reflection you could also check `project.getGradle().getGradleVersion()`

My plugin compiles against the Gradle 2.9 API (the minimum version that we support). I don't want to compile against 3.x and support 2.x as the risk of using Gradle 3-specific API is too high. In short, I think I have no choice but to use reflection.

> Making deprecations a "breaking change" would seriously lock down our options and not really improve the situation.

Given the warning that's shown to users (rather than just at compile time to developers), they're already being seen as a "breaking" change. That's why our users are complaining about them.

> or drop support for the previous major version immediately

We're stuck supported Gradle 2.x (2.9+) and 3.x for Spring Boot 1.5 as that's what we've publicly stated. For Spring Boot 2.0, I think we'll need to reconsider this. Most likely we'll only support a single major version of Gradle and fail hard if anyone tries to do otherwise. That will annoy our users (anecdotally we have a broad mixture of Gradle 2.x and 3.x being used with 1.5), but the ongoing problems with plugins supporting multiple major versions of Gradle makes me think it's the sensible thing for us to do.

**Comment by oehme** [comment]
> Plugin authors can realise that things are deprecated by compiling against the latest API and failing on warnings.

> My plugin compiles against the Gradle 2.9 API (the minimum version that we support). I don't want to compile against 3.x

We have a problem here :) This is exactly why we are showing these warnings at runtime. If we don't tell the users, then the authors often don't know about the deprecation until their plugin starts failing with the next major release.

> I don't want to compile against 3.x and support 2.x as the risk of using Gradle 3-specific API is too high. In short, I think I have no choice but to use reflection.

TestKit allows you to test against multiple Gradle versions, making sure that you are still compatible. I'd encourage you to compile against the latest and ensure compatibility throught tests.

> For Spring Boot 2.0, I think we'll need to reconsider this. Most likely we'll only support a single major version of Gradle and fail hard if anyone tries to do otherwise. That will annoy our users (anecdotally we have a broad mixture of Gradle 2.x and 3.x being used with 1.5), but the ongoing problems with plugins supporting multiple major versions of Gradle makes me think it's the sensible thing for us to do.

This is of course up to you, but this will annoy a lot of your users. It would also slow down adoption of new Gradle releases, which is bad for the whole ecosystem. If you only upgrade your API usage on major version boundaries, then you will likely experience more painful upgrades (because you'll have to handle all deprecations at once), thus further delaying a version that works with a new Gradle release. I think the current early warning system provides a much smoother path.

**Comment by eriwen** [comment]
@wilkinsona Regarding https://github.com/spring-projects/spring-boot/issues/8154, I saw 3.3 in the issue and forgot your version support. I'm glad to continue discussion here.

> The alternative is that every plugin developer who wants to support 2.x and 3.x has to impact the maintainability of their codebase by introducing reflective hacks. As I said above, I think that's an inefficient way to tackle the problem and I would much rather see Gradle take care of this for everyone.

I agree that this sucks. Of course we want an API that can be fully backward-compatible. We're not there yet and it's going to take quite awhile because we'll have to introduce fully breaking changes in the interest of the greater good (in this case, much faster build times). In time the API will anneal and the likelihood of encountering cross-version compatibility problems will decrease.

> Most likely we'll only support a single major version of Gradle and fail hard if anyone tries to do otherwise.

Only supporting a single major version is understandable, but it will somewhat tie your major release schedule to Gradle's as long as there are users providing pressure.

I expect that the number of discussions with disappointed users will be about the same if you fail hard or not. I'd encourage you to allow users to opt-in to higher versions of Gradle with explicit documentation about supported versions, lest we both risk users moving on to other technologies with the latest and greatest features. Further, this will allow users to provide valuable early feedback as they'd try out Spring Boot 2.x with the latest versions of Gradle.

> The problem for me really boils down to how that deprecation warning is being shown. Users are being told about a deprecation that is (largely) outside of their control to fix.

Hmm.. I think there are some things we can do here in the future regarding the way we surface warnings. This is a tough thing to balance. I have some ideas I'll share with the team, but they might be too crazy so I don't want to promise anything at this moment.

**Comment by huxi** [comment]
`ProjectDependency.getProjectConfiguration()` could wrap the remainder of the method in a `Runnable` and then execute it with `SingleMessageLogger.whileDisabled(runnable)` so enclosed warnings are omitted.

This is just meant as a suggestion of how to get rid of this very specific issue in a pragmatic way.

**Comment by adammurdoch** [comment]
It seems there are a bunch of things we can do to make this work better, in general.

The deprecation warnings could be presented in a better way, to separate the things that a plugin author cares about, such as 'you should not be using this particular API as it's broken' or 'this API that you're using will stop working in Gradle 5.0', from the things that a build user cares about, such as 'this plugin won't work with Gradle 5.0'.

We might offer ways to say 'I know' to these warnings. For example, a plugin author might declare that they are targeting Gradle 3.x, in which case we might not complain at runtime to a particular deprecation usage or might dynamically mix in the relevant APIs, as we've started doing in other scenarios. This information (here's the Gradle versions this plugin version was built for/works on) could also be presented to users in the plugin portal and we might also validate this at plugin apply time.

From the build user's perspective, we might move the deprecation warnings at runtime to some 'build inspection' capability that the build user can explicitly perform to say 'tell me about problems with my build'.

For deeper API changes, I think offering the capability to have different variants of a plugin that target specific Gradle versions would be useful.

In this specific case, the deprecated API was _broken_ and has been for a long time. You can't query the data from a dependency declaration and infer what the result will be in isolation. You need to take into account conflicts and conflict resolution, dependency substitution rules, composite builds, component selection rules, and so on. In other words, you need to use the _output_ of dependency resolution, not the _input_. These things have been around for a long time in various forms and this API has been broken for just as long.

**Comment by eriwen** [comment]
@adammurdoch I filed #1372 based on some of your response
